streams.ml

```diff
@@ -0,0 +1,65 @@
+open Effect
+open Printf
+
+type 'a stream_state = {
+  mutex : Mutex.t;
+  capacity : int;
+  item_q : ('a, exn) Result.t Queue.t;
+  readers : 'a Sched.resumer Queue.t;
+  writers : unit Sched.resumer Queue.t;
+}
+
+type 'a t = 'a stream_state
+
+let create capacity =
+  {
+    mutex = Mutex.create ();
+    capacity;
+    item_q = Queue.create ();
+    readers = Queue.create ();
+    writers = Queue.create ();
+  }
+
+let add stream item =
+  Mutex.lock stream.mutex;
+  match Queue.pop stream.readers with
+  (* Resume the resumer with value*)
+  | reader_resumer ->
+      if reader_resumer (Ok item) then printf "Successfully resumed"
+      else printf "Resumer was cancelled";
```

Comment:
Hmm... It seems to me that the logic around cancellation here
1. leaks readers, because cancelled readers are not eagerly removed, and
2. discards items silently by giving them to cancelled readers instead of potential live readers in the queue.